crates/provider-sdk/src/provider.rs

```diff
@@ -81,6 +146,543 @@ macro_rules! process_until_quit {
+                    span.record(
+                        "wit_interfaces",
+                        &tracing::field::display(&ld.interfaces.join(",")),
+                    );
+                    span.record("link_name", &tracing::field::display(&ld.name));
+                    let (tx, rx) = oneshot::channel();
+                    if let Err(err) = link_put_tx.send((ld, tx)).await {
+                        error!(%err, "failed to send link put request");
+                        continue;
+                    }
+                    if let Err(err) = rx.await {
+                        error!(%err, "failed to await link_put")
+                    }
+                }
+                Err(err) => {
+                    error!(%err, "received invalid link def data on message");
+                }
+            }
+        });
+    });
+    Ok(link_put_rx)
+}
+
+async fn subscribe_link_del(
+    nats: Arc<async_nats::Client>,
+    mut quit: broadcast::Receiver<()>,
+    lattice: &str,
+    provider_key: &str,
+) -> ProviderInitResult<mpsc::Receiver<(InterfaceLinkDefinition, oneshot::Sender<()>)>> {
+    let subject = link_del_subject(lattice, provider_key).to_subject();
+    debug!(%subject, "subscribing for link del");
+    let mut sub = nats.subscribe(subject.clone()).await?;
+    let (link_del_tx, link_del_rx) = mpsc::channel(1);
+    let span = tracing::trace_span!("subscribe_link_del", %subject);
+    spawn(
+        async move {
+            process_until_quit!(sub, quit, msg, {
+                if let Ok(ld) = serde_json::from_slice::<InterfaceLinkDefinition>(&msg.payload) {
+                    let (tx, rx) = oneshot::channel();
+                    if let Err(err) = link_del_tx.send((ld, tx)).await {
+                        error!(%err, "failed to send link del request");
+                        continue;
+                    }
+                    if let Err(err) = rx.await {
+                        error!(%err, "failed to await link_del")
+                    }
+                }
+            });
+        }
+        .instrument(span),
+    );
+    Ok(link_del_rx)
+}
+
+pub struct ProviderCommands {
+    pub health: mpsc::Receiver<(HealthCheckRequest, oneshot::Sender<HealthCheckResponse>)>,
+    pub shutdown: mpsc::Receiver<oneshot::Sender<()>>,
+    pub link_put: mpsc::Receiver<(InterfaceLinkDefinition, oneshot::Sender<()>)>,
+    pub link_del: mpsc::Receiver<(InterfaceLinkDefinition, oneshot::Sender<()>)>,
+}
+
+/// State of provider initialization
+pub struct ProviderInitState {
+    pub nats: Arc<async_nats::Client>,
+    pub quit_rx: broadcast::Receiver<()>,
+    pub quit_tx: broadcast::Sender<()>,
+    pub host_id: String,
+    pub lattice_rpc_prefix: String,
+    pub link_name: String,
+    pub provider_key: String,
+    pub link_definitions: Vec<InterfaceLinkDefinition>,
+    pub commands: ProviderCommands,
+}
+
+#[instrument]
+pub async fn init_provider(name: &str) -> ProviderInitResult<ProviderInitState> {
```

Comment:
Does this need to be public? If so, we should document it -- if not, then let's remove `pub`